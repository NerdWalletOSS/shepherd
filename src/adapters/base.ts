export interface IRepo {
  [key: string]: any;
}

export interface IssueTracker {
  [key: string]: any;
}

export interface IEnvironmentVariables {
  [key: string]: string;
}

export type RetryMethod = (opts: number) => any;

interface IRepoAdapter {
  getCandidateRepos(onRetry: RetryMethod): Promise<IRepo[]>;

  parseRepo(repo: string): IRepo;

  reposEqual(repo1: IRepo, repo2: IRepo): boolean;

  stringifyRepo(repo: IRepo): string;

  mapRepoAfterCheckout(repo: Readonly<IRepo>): Promise<IRepo>;

  checkoutRepo(repo: IRepo): Promise<void>;

  resetChangedFiles(repo: IRepo): Promise<void>;

  resetRepoBeforeApply(repo: IRepo, force: boolean): Promise<void>;

  commitRepo(repo: IRepo): Promise<void>;

  pushRepo(repo: IRepo, force: boolean): Promise<void>;

  createPullRequest(repo: IRepo, message: string, upstreamOwner: string): Promise<void>;

  getPullRequestStatus(repo: IRepo): Promise<string[]>;

  getRepoDir(repo: IRepo): string;

  getDataDir(repo: IRepo): string;

  getBaseBranch(repo: IRepo): string;

  getEnvironmentVariables(repo: IRepo): Promise<IEnvironmentVariables>;

  createIssue(repo: IRepo): Promise<string>;

  updateIssue(repo: IRepo, issueNumber: number): Promise<void>;
}

export default IRepoAdapter;
