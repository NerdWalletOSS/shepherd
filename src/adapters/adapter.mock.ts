import IRepoAdapter, { IEnvironmentVariables, IRepo } from './base';

const mockAdapter: IRepoAdapter = {
  getCandidateRepos: jest.fn() as unknown as (onRetry: jest.Mock) => Promise<IRepo[]>,
  parseRepo: jest.fn() as unknown as (repo: string) => IRepo,
  reposEqual: jest.fn() as unknown as (repo1: IRepo, repo2: IRepo) => boolean,
  stringifyRepo: jest.fn() as unknown as (repo: IRepo) => string,
  mapRepoAfterCheckout: jest.fn() as unknown as (repo: Readonly<IRepo>) => Promise<IRepo>,
  checkoutRepo: jest.fn() as unknown as (repo: IRepo) => Promise<void>,
  resetChangedFiles: jest.fn() as unknown as (repo: IRepo) => Promise<void>,
  resetRepoBeforeApply: jest.fn() as unknown as (repo: IRepo, force: boolean) => Promise<void>,
  commitRepo: jest.fn() as unknown as (repo: IRepo) => Promise<void>,
  pushRepo: jest.fn() as unknown as (repo: IRepo, force: boolean) => Promise<void>,
  createPullRequest: jest.fn() as unknown as (
    repo: IRepo,
    message: string,
    upstreamOwner: string
  ) => Promise<void>,
  getPullRequestStatus: jest.fn() as unknown as (repo: IRepo) => Promise<string[]>,
  getRepoDir: jest.fn() as unknown as (repo: IRepo) => string,
  getDataDir: jest.fn() as unknown as (repo: IRepo) => string,
  getBaseBranch: jest.fn() as unknown as (repo: IRepo) => string,
  getEnvironmentVariables: jest.fn() as unknown as (repo: IRepo) => Promise<IEnvironmentVariables>,
  createIssue: jest.fn() as unknown as (repo: IRepo) => Promise<string>,
  updateIssue: jest.fn() as unknown as (repo: IRepo) => Promise<void>,
};

export default mockAdapter;
