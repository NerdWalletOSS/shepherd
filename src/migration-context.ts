import IRepoAdapter, { IRepo } from './adapters/base.js';
import { ILogger } from './logger/index.js';
import { IMigrationSpec } from './util/migration-spec.js';

export interface IShepherdInfo {
  workingDirectory: string;
}

export interface IMigrationInfo {
  spec: IMigrationSpec;
  migrationDirectory: string;
  workingDirectory: string;
  repos: IRepo[] | null;
  upstreamOwner: string;
  selectedRepos?: IRepo[];
}

export interface IMigrationContext {
  shepherd: IShepherdInfo;
  migration: IMigrationInfo;
  adapter: IRepoAdapter;
  logger: ILogger;
}

export enum state {
  open,
  closed,
}

export enum state_reason {
  completed,
  not_planned,
  reopened,
  null,
}
