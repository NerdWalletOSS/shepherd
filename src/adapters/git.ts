import fs from 'fs-extra';
import { simpleGit } from 'simple-git';

import { IMigrationContext } from '../migration-context.js';
import IRepoAdapter, { IEnvironmentVariables, IRepo, RetryMethod } from './base.js';

abstract class GitAdapter implements IRepoAdapter {
  protected migrationContext: IMigrationContext;
  protected branchName: string;
  constructor(migrationContext: IMigrationContext) {
    this.migrationContext = migrationContext;
    this.branchName = migrationContext.migration.spec.id;
  }

  public abstract getCandidateRepos(onRetry: RetryMethod): Promise<IRepo[]>;

  public abstract parseRepo(repo: string): IRepo;

  public abstract reposEqual(repo1: IRepo, repo2: IRepo): boolean;

  public abstract stringifyRepo(repo: IRepo): string;

  public abstract getRepoDir(repo: IRepo): string;

  public abstract getDataDir(repo: IRepo): string;

  public abstract mapRepoAfterCheckout(repo: Readonly<IRepo>): Promise<IRepo>;

  public abstract resetRepoBeforeApply(repo: IRepo, force: boolean): Promise<void>;

  public async checkoutRepo(repo: IRepo): Promise<void> {
    const repoPath = this.getRepositoryUrl(repo);
    const localPath = this.getRepoDir(repo);
    const localPathExistsAndIsRepo =
      (await fs.pathExists(localPath)) && (await this.git(repo).checkIsRepo());
    if (localPathExistsAndIsRepo) {
      // Repo already exists; just fetch
      await this.git(repo).fetch('origin');
    } else {
      const git = simpleGit();
      await git.clone(repoPath, localPath, ['--depth', '1']);
    }

    // We'll immediately create and switch to a new branch
    try {
      await this.git(repo).checkout(['-b', this.branchName, `origin/${this.branchName}`]);
    } catch {
      try {
        await this.git(repo).checkoutLocalBranch(this.branchName);
      } catch {
        // This branch probably already exists; we'll just switch to it
        // to make sure we're on the right branch for the commit phase
        await this.git(repo).checkout(this.branchName);
      }
    }
  }

  public async commitRepo(repo: IRepo, noVerify:boolean): Promise<void> {
    const {
      migration: { spec },
    } = this.migrationContext;
    const options = noVerify ? ['--no-verify'] : undefined;
    await this.git(repo).add('.');
    await this.git(repo).commit(`${spec.title} [shepherd]`, options);
  }

  public async resetChangedFiles(repo: IRepo): Promise<void> {
    await this.git(repo).reset(['--hard']);
    await this.git(repo).clean('f', ['-d']);
  }

  public async pushRepo(repo: IRepo, force: boolean, noVerify: boolean): Promise<void> {
    const options = []
    if (force) { options.push('--force') }
    if (noVerify) { options.push('--no-verify') }
    await this.git(repo).push('origin', 'HEAD', options);
  }

  public abstract createPullRequest(
    repo: IRepo,
    message: string,
    upstreamOwner: string
  ): Promise<void>;

  public abstract getPullRequestStatus(repo: IRepo): Promise<string[]>;

  public abstract getBaseBranch(repo: IRepo): string;

  public async getEnvironmentVariables(repo: IRepo): Promise<IEnvironmentVariables> {
    const revision = await this.git(repo).revparse(['HEAD']);

    return {
      SHEPHERD_GIT_REVISION: revision,
    };
  }

  protected abstract getRepositoryUrl(repo: IRepo): string;

  protected git(repo: IRepo): any {
    return simpleGit(this.getRepoDir(repo));
  }

  protected isShepherdCommitMessage(message: string): boolean {
    return message.indexOf('Shepherd: ') === 0 || message.indexOf('[shepherd]') !== -1;
  }
}
export default GitAdapter;
