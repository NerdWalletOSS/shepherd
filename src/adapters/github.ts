import chalk from 'chalk';
import _ from 'lodash';
import path from 'path';

import { IMigrationContext } from '../migration-context.js';
import { IEnvironmentVariables, IRepo } from './base.js';
import GitAdapter from './git.js';
import GithubService from '../services/github.js';

enum SafetyStatus {
  Success,
  PullRequestExisted,
  NonShepherdCommits,
}

class GithubAdapter extends GitAdapter {
  private githubService: GithubService;

  constructor(migrationContext: IMigrationContext, githubService: GithubService) {
    super(migrationContext);
    this.migrationContext = migrationContext;
    this.githubService = githubService;
  }

  public async getCandidateRepos(): Promise<IRepo[]> {
    const {
      org,
      search_type = 'code',
      search_query,
    } = this.migrationContext.migration.spec.adapter;
    let repoNames: string[];

    // list all of an orgs active repos
    if (org) {
      if (search_query) {
        throw new Error('Cannot use both "org" and "search_query" in GitHub adapter. Pick one.');
      }

      repoNames = await this.githubService.getActiveReposForOrg({ org });
    } else {
      repoNames = await this.githubService.getActiveReposForSearchTypeAndQuery({
        search_type,
        search_query,
      });
    }

    return _.uniq(repoNames).map((r) => this.parseRepo(r));
  }

  public async mapRepoAfterCheckout(repo: Readonly<IRepo>): Promise<IRepo> {
    const { owner, name } = repo;
    const defaultBranch = await this.githubService.getDefaultBranchForRepo({
      owner,
      repo: name,
    });

    return {
      ...repo,
      defaultBranch,
    };
  }

  public parseRepo(repo: string): IRepo {
    const [owner, name] = repo.split('/');
    if (!owner || !name) {
      throw new Error(`Could not parse repo "${repo}"`);
    }
    return { owner, name };
  }

  public reposEqual(repo1: IRepo, repo2: IRepo): boolean {
    // GitHub ignores case when differentiating organizations and repos, so we
    // can safely use a case-insensitive compare to make things slightly easier
    // for users who might be using a case-insensitive name on the command line.
    return (
      repo1.owner.toLowerCase() === repo2.owner.toLowerCase() &&
      repo1.name.toLowerCase() === repo2.name.toLowerCase()
    );
  }

  public stringifyRepo({ owner, name }: IRepo): string {
    return `${owner}/${name}`;
  }

  public async resetRepoBeforeApply(repo: IRepo, force: boolean) {
    const { defaultBranch } = repo;
    // First, get any changes from the remote
    // --prune will ensure that any remote branch deletes are reflected here
    await this.git(repo).fetch(['origin', '--prune']);

    if (!force) {
      const safetyStatus = await this.checkActionSafety(repo);
      if (safetyStatus === SafetyStatus.PullRequestExisted) {
        throw new Error(
          'Remote branch did not exist, but a pull request does or did; try with --force-reset-branch?'
        );
      } else if (safetyStatus === SafetyStatus.NonShepherdCommits) {
        throw new Error(
          'Found non-Shepherd commits on remote branch; try with --force-reset-branch?'
        );
      }
    }

    // If we got this far, we can go ahead and reset to the default branch
    await this.git(repo).reset(['--hard', `origin/${defaultBranch}`]);
  }

  public async pushRepo(repo: IRepo, force: boolean, noVerify: boolean): Promise<void> {
    let shouldForce = false;

    // First, get any changes from the remote
    // --prune will ensure that any remote branch deletes are reflected here
    await this.git(repo).fetch(['origin', '--prune']);

    if (!force) {
      const safetyStatus = await this.checkActionSafety(repo);
      if (safetyStatus === SafetyStatus.PullRequestExisted) {
        throw new Error(
          'Remote branch did not exist, but a pull request does or did; try with --force?'
        );
      } else if (safetyStatus === SafetyStatus.NonShepherdCommits) {
        throw new Error('Found non-Shepherd commits on remote branch; try with --force?');
      }

      // If we get to here, it's safe to force-push to this branch
      shouldForce = true;
    }

    await super.pushRepo(repo, force || shouldForce, noVerify);
  }

  public async createPullRequest(
    repo: IRepo,
    message: string,
    upstreamOwner: string
  ): Promise<void> {
    const {
      migration: { spec },
    } = this.migrationContext;
    const { owner, name, defaultBranch } = repo;

    let baseOwner = owner;

    if (upstreamOwner) {
      baseOwner = upstreamOwner;
    }

    // Let's check if a PR already exists
    const pullRequests = await this.githubService.listPullRequests({
      owner,
      repo: name,
      head: `${owner}:${this.branchName}`,
    });

    if (pullRequests && pullRequests.length) {
      const pullRequest = pullRequests[0];

      if (pullRequest.state === 'open') {
        // A pull request exists and is open, let's update it
        await this.githubService.updatePullRequest({
          owner: baseOwner,
          repo: name,
          pull_number: pullRequest.number,
          title: spec.title,
          body: message,
        });
      } else {
        // A pull request exists but it was already closed - don't update it
        // TODO proper status reporting without errors
        throw new Error('Could not update pull request; it was already closed');
      }
    } else {
      // No PR yet - we have to create it

      await this.githubService.createPullRequest({
        owner: baseOwner,
        repo: name,
        head: `${owner}:${this.branchName}`,
        base: defaultBranch,
        title: spec.title,
        body: message,
      });
    }
  }

  public async getPullRequestStatus(repo: IRepo): Promise<string[]> {
    const { owner, name } = repo;
    const status: string[] = [];

    // First, check for a pull request
    const pullRequests = await this.githubService.listPullRequests({
      owner,
      repo: name,
      head: `${owner}:${this.branchName}`,
      state: 'all',
    });

    if (pullRequests && pullRequests.length) {
      // GitHub's API is weird - you need a second query to get information about mergeability
      const { data: pullRequest } = await this.githubService.getPullRequest({
        owner,
        repo: name,
        pull_number: pullRequests[0].number,
      });

      status.push(`PR #${pullRequest.number} [${pullRequest.html_url}]`);
      if (pullRequest.merged_at) {
        status.push(chalk.magenta(`PR was merged at ${pullRequest.merged_at}`));
      } else if (pullRequest.mergeable && pullRequest.mergeable_state === 'clean') {
        status.push(chalk.green('PR is mergeable!'));
      } else {
        status.push(chalk.red('PR is not mergeable'));
        // Let's see what's blocking us
        // Sadly, we can only get information about failing status checks, not being blocked
        // by things like required reviews
        const combinedStatus = await this.githubService.getCombinedRefStatus({
          owner,
          repo: name,
          ref: this.branchName,
        });

        const { statuses } = combinedStatus.data;
        const anyPending = statuses.some((s: any) => s.state === 'pending');
        const anyFailing = statuses.some((s: any) => s.state === 'error' || s.state === 'failure');

        const recordStatus = (s: any) =>
          status.push(`${s.context} ${chalk.dim(`- ${s.description}`)}`);

        if (anyPending) {
          status.push(chalk.underline.yellow('Pending status checks'));
          statuses.forEach((s: any) => {
            if (s.state !== 'pending') {
              return;
            }
            recordStatus(s);
          });
        }

        if (anyFailing) {
          status.push(chalk.underline.red('Failing status checks'));
          statuses.forEach((s: any) => {
            if (!(s.state === 'error' || s.state === 'failure')) {
              return;
            }
            recordStatus(s);
          });
        }
      }
    } else {
      try {
        // This will throw an exception if the branch does not exist
        await this.githubService.getBranch({
          owner,
          repo: name,
          branch: this.branchName,
        });
        status.push('No PR exists');
      } catch (e: any) {
        if (e.code === 404) {
          status.push('No branch or PR exists');
        } else {
          throw e;
        }
      }
    }

    return status;
  }

  public getRepoDir(repo: IRepo): string {
    return path.join(
      this.migrationContext.migration.workingDirectory,
      'repos',
      repo.owner,
      repo.name
    );
  }

  public getDataDir(repo: IRepo): string {
    return path.join(
      this.migrationContext.migration.workingDirectory,
      'data',
      repo.owner,
      repo.name
    );
  }

  public getOwnerName(owner: string): string {
    return owner;
  }

  public getBaseBranch(repo: IRepo): string {
    return repo.defaultBranch;
  }

  public async getEnvironmentVariables(repo: IRepo): Promise<IEnvironmentVariables> {
    const superEnvVars = await super.getEnvironmentVariables(repo);

    return {
      ...superEnvVars,
      SHEPHERD_GITHUB_REPO_OWNER: repo.owner,
      SHEPHERD_GITHUB_REPO_NAME: repo.name,
    };
  }

  protected getRepositoryUrl(repo: IRepo): string {
    const gitHubEnterpriseBaseUrl =
      process.env.SHEPHERD_GITHUB_ENTERPRISE_BASE_URL ?? 'api.github.com';
    const githubProtocol = process.env.SHEPHERD_GITHUB_PROTOCOL ?? 'ssh';

    if (githubProtocol === 'ssh') {
      return `git@${gitHubEnterpriseBaseUrl}:${repo.owner}/${repo.name}.git`;
    } else if (githubProtocol === 'https') {
      return `https://${gitHubEnterpriseBaseUrl}/${repo.owner}/${repo.name}.git`;
    } else {
      throw new Error(`Unknown protocol ${githubProtocol}. Valid values are 'ssh' and 'https'`);
    }
  }

  private async checkActionSafety(repo: IRepo): Promise<SafetyStatus> {
    const { owner, name } = repo;

    // Get all branches and look for the remote branch
    // @ts-ignore (typings are broken)
    const { branches } = await this.git(repo).branch();
    if (branches[`remotes/origin/${this.branchName}`] === undefined) {
      // This remote branch does not exist
      // We need to figure out if that's because a PR was open and
      // subsequently closed, or if it's because we just haven't pushed
      // a branch yet
      const pullRequests = await this.githubService.listPullRequests({
        owner,
        repo: name,
        head: `${owner}:${this.branchName}`,
        state: 'all',
      });

      if (pullRequests && pullRequests.length) {
        // We'll assume that if a remote branch does not exist but a PR
        // does/did, we don't want to apply to this branch
        return SafetyStatus.PullRequestExisted;
      }
    } else {
      // The remote branch exists!
      // We'll get the list of all commits not on main and check if they're
      // all from Shepherd. If they are, it's safe to reset the branch to
      // main.
      const upstreamBranch = `remotes/origin/${this.branchName}`;
      const commits = await this.git(repo).log([`HEAD..${upstreamBranch}`]);
      const allShepherd = commits.all.every(({ message }: any) =>
        this.isShepherdCommitMessage(message)
      );
      if (!allShepherd) {
        // RIP.
        return SafetyStatus.NonShepherdCommits;
      }
    }

    return SafetyStatus.Success;
  }
}

export default GithubAdapter;
