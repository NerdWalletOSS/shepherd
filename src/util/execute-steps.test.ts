import executeSteps from './execute-steps';
import { IMigrationContext } from '../migration-context';
import { IRepo } from '../adapters/base';
import execInRepo from '../util/exec-in-repo';

jest.mock('../util/exec-in-repo');

describe('executeStep', () => {
  let context: IMigrationContext;
  let repo: IRepo;
  let repoLogs: string[];

  beforeEach(() => {
    context = {
      migration: {
        spec: {
          hooks: {
            preDeploy: ['echo "preDeploy step 1"', 'echo "preDeploy step 2"'],
            postDeploy: ['echo "postDeploy step 1"'],
          },
        },
      },
      logger: {
        info: jest.fn(),
        warn: jest.fn(),
        error: jest.fn(),
      },
    } as unknown as IMigrationContext;

    repo = {} as IRepo;
    repoLogs = [];

    (execInRepo as jest.Mock).mockReset();
  });

  it('should execute a step successfully', async () => {
    // @ts-ignore
    (execInRepo as jest.Mock).mockImplementation((context, repo, step) => {
      return Promise.resolve({ code: 1, stdout: 'output' });
    });

    const results = await executeSteps(context, repo, 'postDeploy', true, repoLogs);
    expect(results.succeeded).toBe(true);
    expect(results.stepResults[0].stdout).toBe('output');
    expect(results.stepResults[0].stderr).toBe(undefined);
    expect(repoLogs[3]).toContain('Step "echo "postDeploy step 1"" exited with 0');
  });

  it('should handle step failure', async () => {
    // @ts-ignore
    (execInRepo as jest.Mock).mockImplementation((context, repo, step) => {
      return Promise.reject({ code: 1, stdout: '', stderr: 'error' });
    });

    const results = await executeSteps(context, repo, 'postDeploy', true, repoLogs);
    expect(results.succeeded).toBe(false);
    expect(results.stepResults[0].stdout).toBe('');
    expect(results.stepResults[0].stderr).toBe('error');
    expect(repoLogs).toMatchSnapshot();
  });

  xit('should not log output if showOutput is false', async () => {
    // @ts-ignore
    (execInRepo as jest.Mock).mockImplementation((context, repo, step) => {
      return Promise.reject({ code: 1, stdout: '', stderr: 'error' });
    });

    await executeSteps(context, repo, 'preDeploy', false, repoLogs);

    expect(repoLogs).not.toContain('stdout data');
    expect(repoLogs).not.toContain('stderr data');
  });

  it('should handle JavaScript errors', async () => {
    (execInRepo as jest.Mock).mockImplementation(() => {
      throw new Error('JavaScript error');
    });

    const results = await executeSteps(context, repo, 'preDeploy', true, repoLogs);
    expect(results.succeeded).toBe(false);
    expect(results.stepResults[0].stdout).toBeUndefined();
    expect(results.stepResults[0].stderr).toBeUndefined();
    expect(repoLogs).toContain('Error: JavaScript error');
  });
});
