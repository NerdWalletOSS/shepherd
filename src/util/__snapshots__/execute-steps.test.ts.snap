// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executeStep should handle step failure 1`] = `
[
  "$ echo "postDeploy step 1"",
  "Step "echo "postDeploy step 1"" exited with 1",
]
`;
